bluez/plugins/jolla-wakelock.c

```diff
@@ -0,0 +1,293 @@
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+/* Plugin that implements wakelocks. Only one real wakelock is ever
+   used; API can create several "virtual" wakelocks which are mapped
+   on the real one: 
+
+   - no virtual wakelock active -> real wakelock not active
+   - at least one virtual wakelock active -> real wakelock active
+*/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <errno.h>
+#include <glib.h>
+
+#include "plugin.h"
+#include "log.h"
+#include "wakelock.h"
+
+#define LOCK_FILE	"/sys/power/wake_lock"
+#define UNLOCK_FILE	"/sys/power/wake_unlock"
+#define LOCK_NAME	"bluetoothd"
+#define LOCK_NAME_LEN   10
+
+struct wakelock {
+	gchar *name;
+	gsize acquisitions;
+	gboolean stale;
+};
+
+static GSList *locks = NULL;
+
+static gboolean initialized = FALSE;
+
+static gboolean real_lock_acquired = FALSE;
+
+static int sysfs_write(const gchar *file, const gchar *data_ptr, gsize data_len)
+{
+	int fd = -1;
+	int r;
+
+	fd = open(file, O_WRONLY);
+	if (fd < 0) {
+		error("Failed to open: %s(%d)", strerror(errno), errno);
+		r = -errno;
+		goto out;
+	}
+
+	if (write(fd, data_ptr, data_len) != data_len) {
+		error("Atomic write failed");
+		r = -EIO;
+		goto out;
+	}
+
+	r = 0;
+
+out:
+	if (fd >= 0)
+		close(fd);
+
+	return r;
+}
+
+static int jolla_wakelock_acquire(struct wakelock *lock)
+{
+	int r;
+
+	DBG("lock: %p, name: %s, acquisitions: %u, stale: %s",
+		lock, lock->name, lock->acquisitions,
+		lock->stale ? "yes" : "no");
+
+	if (lock->stale)
+		return -EBADF;
+
+	if (real_lock_acquired)
+		goto done;
+
+	r = sysfs_write(LOCK_FILE, LOCK_NAME, LOCK_NAME_LEN);
+	if (r < 0) {
+		error("Failed to acquire real lock for lock: %p, name %s",
+			lock, lock->name);
+		return r;
+	}
+
+	real_lock_acquired = TRUE;
+	DBG("Real wakelock acquired for lock: %p, name %s", lock, lock->name);
+
+done:
+	lock->acquisitions++;
+	DBG("lock: %p, name %s acquired", lock, lock->name);
+
+	return 0;
+}
+
+static int jolla_wakelock_release(struct wakelock *lock)
+{
+	GSList *l;
+	int r;
+
+	DBG("lock: %p, name: %s, acquisitions: %u, stale: %s",
+		lock, lock->name, lock->acquisitions,
+		lock->stale ? "yes" : "no");
+
+	if (lock->stale)
+		return -EBADF;
+
+	if (!lock->acquisitions) {
+		warn("Attempted to release already released lock: %p", lock);
+		return -EINVAL;
+	}
+
+	lock->acquisitions--;
+	if (lock->acquisitions)
+		goto done;
+
+	/* See if this was the last active lock, only release real lock if so */
+	for (l = locks; l; l = l->next) {
+		struct wakelock *check = l->data;
+		if (check->acquisitions)
+			goto done;
+	}
+
+	r = sysfs_write(UNLOCK_FILE, LOCK_NAME, LOCK_NAME_LEN);
+	if (r < 0) {
+		error("Failed to release real lock for lock: %p, name: %s",
+			lock, lock->name);
+		return r;
```

Comment:
Not sure how real the threat is, but if somebody else removes the "bluetoothd" wakelock, this error path gets hit
and book keeping stays at "have real lock" state incorrectly. Just might be lesser evil to log the error, but not bail out via early return so that the next acquire at least attempts to take the lock again?